### ISSUE TITLE
Fix app module tests for Roboelectric

### ISSUE THREAD

**Issue by rt4914** [oppia/oppia-android#973]
Currently, the `app` module test cases are passing on `Espresso` but they should also pass on `Robolectric`.  Fix these below mentioned files to make sure that the tests pass on `Robolectric`.

You can run robolectric tests by following these steps: https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#running-roboelectric-tests

Also, make sure that you assign yourself one task only at a time, which you can mention in the comments below.

Fixing test cases will need to introduce `TextApplicationComponent`. This can be seen [here](https://github.com/oppia/oppia-android/pull/1680/commits/4729d551b86282ce617e1d53f2762416758785a1).

- [x] `AppVersionActivityTest` (Assigned to @anandwana001 )
- [x] `CompletedStoryListActivityTest` (Assigned to @anandwana001 )
- [x] `FAQListFragmentTest` (Assigned to @Sarthak2601  )
- [x] `FAQSingleActivityTest` (Assigned to @SayantanBanerjee16 )
- [x] `HelpFragmentTest` (Assigned to @SayantanBanerjee16 )
- [x] `MyDownloadsFragmentTest`(Assigned to @anandwana001 )
- [x] `HtmlParserTest` (Assigned to @anandwana001 )
- [x] `StateFragmentTest` (Fixed in #1630)
- [x] `AddProfileActivityTest` (Assigned to @anandwana001 )
- [x] `AdminAuthActivityTest` (Assigned to @anandwana001 )
- [x] `AdminPinActivityTest` (Assigned to @anandwana001 )
- [x] `ProfilePictureActivityTest`(Assigned to @anandwana001 )
- [x] `ProfileProgressFragmentTest` (Assigned to @veena14cs )
- [x] `BindableAdapterTest` (Assigned to @anandwana001 )
- [x] `SplashActivityTest` (Fixed in #1397)
- [x] `StoryActivityTest` (Assigned to @anandwana001 )
- [x] `StoryFragmentTest` (Assigned to @anandwana001 )
- [x] `QuestionPlayerActivityTest` (Assigned to @anandwana001)
- [x] `TopicTestActivityForStoryTest`(Assigned to @anandwana001 )
- [x] `RevisionCardFragmentTest` (Assigned to @anandwana001 )
- [x] `WalkthroughWelcomeFragmentTest`(Assigned to @veena14cs )
- [x] `WalkthroughTopicListFragmentTest`(Assigned to @anandwana001 )
- [x] `WalkthroughFinalFragmentTest` (Assigned to @anandwana001 )
- [x] `WalkthroughActivityTest`(Assigned to @anandwana001 )

**Comment by abhinavraj23** [comment]
@rt4914 please assign me one of these tasks.

**Comment by SayantanBanerjee16** [comment]
@rt4914 Can you assign me some of this task. Also, this task means that we would be running the previously written tests in Roboelectric, or writing new tests based on previous tests?

**Comment by rt4914** [comment]
> @rt4914 Can you assign me some of this task. Also, this task means that we would be running the previously written tests in Roboelectric, or writing new tests based on previous tests?

You can just run the previously written test cases in Roboelectic and which every test fails, fix them and make sure that it passes on Espresso as well as Roboelectric

**Comment by SayantanBanerjee16** [comment]
@rt4914 So I followed the steps given in the link above and all the tests in `AppVersionActivityTest.kt ` passed. Now can you please assign me files 4,5,6 in the list above. Thanks :)

![Screenshot (101)](https://user-images.githubusercontent.com/43978143/79548040-72655900-80b2-11ea-901f-30f491dadacd.png)

**Comment by virajpatel2411** [comment]
@rt4914 Assign me any one of these tasks. Thanks.

**Comment by SayantanBanerjee16** [comment]
@rt4914 `FAQSingleActivityTest` Roboelectric tests passed.
I am assigning myself the next two files in the list.

![Screenshot (105)](https://user-images.githubusercontent.com/43978143/79736468-7514b800-8317-11ea-8586-590e0683748e.png)

**Comment by virajpatel2411** [comment]
@rt4914 In `CompletedStoryListActivityTest`, 8 out of 9 tests are failing for Roboelectric and all of then are passing in Espresso. The failed tests show the error of  PerformException, so maybe there seems to be an issue with the `perform()` method. So, what do you suggest to change, because I am not getting what is wrong with the method if all the tests are passing for Espresso.

**Comment by abhinavraj23** [comment]
@rt4914 for the tests concerning dialog-box opening or navigation-bar opening, it would not work with roboelectric in the same way as it works in espresso, thus for these we may need to write separate tests, I have been able to successfully write tests for dialogBoxes in roboelectric but those don't work in espresso, thus we may need to write separately for these tests

**Comment by Sarthak2601** [comment]
All Robolectric tests pass in `FAQListFragmentTest`

![FAQListFragmentRoboelectricTest](https://user-images.githubusercontent.com/53938155/79788949-1119e000-8367-11ea-9b68-b0c5998a6b69.png)

**Comment by rt4914** [comment]
> @rt4914 for the tests concerning dialog-box opening or navigation-bar opening, it would not work with roboelectric in the same way as it works in espresso, thus for these we may need to write separate tests, I have been able to successfully write tests for dialogBoxes in roboelectric but those don't work in espresso, thus we may need to write separately for these tests

@abhinavraj23 Its write to add new test cases as per requirement but each and every test case should pass on both espresso and roboelectric. So you can just create a PR and we can figure out from that.

**Comment by rt4914** [comment]
> @rt4914 In `CompletedStoryListActivityTest`, 8 out of 9 tests are failing for Roboelectric and all of then are passing in Espresso. The failed tests show the error of PerformException, so maybe there seems to be an issue with the `perform()` method. So, what do you suggest to change, because I am not getting what is wrong with the method if all the tests are passing for Espresso.

@virajpatel2411 It can happen that tests pass on espresso and fails on roboelectic and that's why we have this issue to track everything. Have you tried searching for solutions for those errors on internet?

**Comment by virajpatel2411** [comment]
> > @rt4914 In `CompletedStoryListActivityTest`, 8 out of 9 tests are failing for Roboelectric and all of then are passing in Espresso. The failed tests show the error of PerformException, so maybe there seems to be an issue with the `perform()` method. So, what do you suggest to change, because I am not getting what is wrong with the method if all the tests are passing for Espresso.
> 
> @virajpatel2411 It can happen that tests pass on espresso and fails on roboelectic and that's why we have this issue to track everything. Have you tried searching for solutions for those errors on internet?

Yes, but I didn't find anything helpful. Imk, there resides some issue with `perform()` method as in every  test case in Roboelectric the `PerformException` is thrown. Now, what should I do?

**Comment by rt4914** [comment]
> > > @rt4914 In `CompletedStoryListActivityTest`, 8 out of 9 tests are failing for Roboelectric and all of then are passing in Espresso. The failed tests show the error of PerformException, so maybe there seems to be an issue with the `perform()` method. So, what do you suggest to change, because I am not getting what is wrong with the method if all the tests are passing for Espresso.
> > 
> > 
> > @virajpatel2411 It can happen that tests pass on espresso and fails on roboelectic and that's why we have this issue to track everything. Have you tried searching for solutions for those errors on internet?
> 
> Yes, but I didn't find anything helpful. Imk, there resides some issue with `perform()` method as in every test case in Roboelectric the `PerformException` is thrown. Now, what should I do?

@virajpatel2411 You can either prepare a document of what all things you have tried and the errors you have faced so that we can have a look. 
2. You can pick a new issue to work on from this list.

But the main point is that you will need to give information in detail so that we can have a look at the problem.

**Comment by virajpatel2411** [comment]
> > > > @rt4914 In `CompletedStoryListActivityTest`, 8 out of 9 tests are failing for Roboelectric and all of then are passing in Espresso. The failed tests show the error of PerformException, so maybe there seems to be an issue with the `perform()` method. So, what do you suggest to change, because I am not getting what is wrong with the method if all the tests are passing for Espresso.
> > > 
> > > 
> > > @virajpatel2411 It can happen that tests pass on espresso and fails on roboelectic and that's why we have this issue to track everything. Have you tried searching for solutions for those errors on internet?
> > 
> > 
> > Yes, but I didn't find anything helpful. Imk, there resides some issue with `perform()` method as in every test case in Roboelectric the `PerformException` is thrown. Now, what should I do?
> 
> @virajpatel2411 You can either prepare a document of what all things you have tried and the errors you have faced so that we can have a look.
> 2. You can pick a new issue to work on from this list.
> 
> But the main point is that you will need to give information in detail so that we can have a look at the problem.

Okay...Will share the details once I have prepared it. Thanks.

**Comment by virajpatel2411** [comment]
@rt4914 https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing. 
This is the link of the document I prepared. Hope it helps you understand the problem faced.

**Comment by virajpatel2411** [comment]
> @rt4914 https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing.
> This is the link of the document I prepared. Hope it helps you understand the problem faced.

@rt4914 Any updates for further work?

**Comment by rt4914** [comment]
> @rt4914 https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing.
> This is the link of the document I prepared. Hope it helps you understand the problem faced.

@veena14cs PTAL at this issue.

**Comment by virajpatel2411** [comment]
> > @rt4914 https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing.
> > This is the link of the document I prepared. Hope it helps you understand the problem faced.
> 
> @veena14cs PTAL at this issue.

@veena14cs This is the link to document.
https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing
Thanks.

**Comment by virajpatel2411** [comment]
> > > @rt4914 https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing.
> > > This is the link of the document I prepared. Hope it helps you understand the problem faced.
> > 
> > 
> > @veena14cs PTAL at this issue.
> 
> @veena14cs This is the link to document.
> https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing
> Thanks.

@rt4914 @veena14cs Any updates?

**Comment by veena14cs** [comment]
> https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing

@virajpatel2411 I checked the document will figure this out, till then you can pick out some new issue and start working on it. Once we figure out, I will update you on this.

**Comment by virajpatel2411** [comment]
> > https://docs.google.com/document/d/12vTSL8eW9LMtPmxSGFnWvmQfJumJ4BOdBduqeUm8vmw/edit?usp=sharing
> 
> @virajpatel2411 I checked the document will figure this out, till then you can pick out some new issue and start working on it. Once we figure out, I will update you on this.

@veena14cs Okay.

**Comment by anandwana001** [comment]
@virajpatel2411 I tried `CompletedStoryListActivityTest` running `Robolectric` test and yes, it passed only 1 out of 9 in my case. I annotated `@LooperMode(PAUSED)` annotation on the top of the class and it passed all the test cases under `Robolectric Testing` 

added `androidTestImplementation 'org.robolectric:annotations:4.3.1'`

[Documentation of Robolectric Looper](http://robolectric.org/blog/2019/06/04/paused-looper/)

[Reference to this Robolectric Issue](https://github.com/robolectric/robolectric/issues/5385)

[LooperMode class](https://github.com/robolectric/robolectric/blob/master/annotations/src/main/java/org/robolectric/annotation/LooperMode.java)

Robolectric Result
<img width="1440" alt="Screenshot 2020-05-15 at 22 36 59" src="https://user-images.githubusercontent.com/25057618/82077269-c2a8f880-96fc-11ea-8b92-db14c8c3d92f.png">

Espresso Result
<img width="1440" alt="Screenshot 2020-05-15 at 22 36 24" src="https://user-images.githubusercontent.com/25057618/82077429-0a2f8480-96fd-11ea-82b4-095eaf5c32e5.png">

**Comment by virajpatel2411** [comment]
@anandwana001 Okay. So, now is there anything to do from my side or the issue now is resolved?

**Comment by anandwana001** [comment]
> @anandwana001 Okay. So, now is there anything to do from my side or the issue now is resolved?

I think you can complete this issue by making a PR if you haven't created yet and complete this issue. Running this test file in espresso fails one of the tests which you can look at and with this it resolves this issue with `CompletedStoryListActivityTest`. Also, you can have a look at, adding this new dependency `androidTestImplementation 'org.robolectric:annotations:4.3.1'` is not bringing any new conflicts in the project.

**Comment by virajpatel2411** [comment]
> > @anandwana001 Okay. So, now is there anything to do from my side or the issue now is resolved?
> 
> I think you can complete this issue by making a PR if you haven't created yet and complete this issue. Running this test file in espresso fails one of the tests which you can look at and with this it resolves this issue with `CompletedStoryListActivityTest`. Also, you can have a look at, adding this new dependency `androidTestImplementation 'org.robolectric:annotations:4.3.1'` is not bringing any new conflicts in the project. 

@anandwana001 Okay. Will make a PR soon. Thanks.

**Comment by anandwana001** [comment]
As far as I saw, most of the errors from most of the above-mentioned files will get resolved using annotation `@LooperMode(LooperMode.Mode.PAUSED)` and `shadowOf(Looper.getMainLooper()).idle()`.
Updating dependency to `robolectric-4.3.1` might resolves lot of tests as `Robolectric` itself doing a lot of rewriting with their current ongoing issues.

**Comment by anandwana001** [comment]
@rt4914  We can remove `WalkthroughWelcomeFragmentTest` from the list as all the test cases are passing in `Espresso` as well as `Robolectric`

<img width="1440" alt="Screenshot 2020-05-16 at 09 44 41" src="https://user-images.githubusercontent.com/25057618/82110211-0a11a200-975a-11ea-9943-6727bd96ff87.png">

**Comment by anandwana001** [comment]
@rt4914  Tried one test `testWalkthroughFragment_increaseProgress_worksCorrectly` from `WalkthroughActivityTest` file and to make it pass on both `espresso` and `robolectric`, I had to make some modification in UI files.
Earlier error was `at least 90 per cent of the view should be visible` and to resolve this, making `ScrollView` as a parent in XML file and using `scrollTo()` while performing `click()` in the test.

Does modifying UI files in order to make the test case pass is valid for this issue? or we had to find some other workarounds?

**Comment by rt4914** [comment]
> @rt4914 Tried one test `testWalkthroughFragment_increaseProgress_worksCorrectly` from `WalkthroughActivityTest` file and to make it pass on both `espresso` and `robolectric`, I had to make some modification in UI files.
> Earlier error was `at least 90 per cent of the view should be visible` and to resolve this, making `ScrollView` as a parent in XML file and using `scrollTo()` while performing `click()` in the test.
> 
> Does modifying UI files in order to make the test case pass is valid for this issue? or we had to find some other workarounds?

@anandwana001 Yes it is valid for this case. We can change UI only in the case when we are actually making UI better and by adding scrolling you are making it better to use. Still, for final confirmation we will need that as a PR to make sure that it does not affect that feature.

**Comment by anandwana001** [comment]
@veena14cs `MyDownloadsFragmentTest`, `ProfilePictureActivityTest` and `InputInteractionViewTestActivityTest` is good. Could we mark them done on the list? 

`MyDownloadsFragmentTest`
Robolectric Test Passes
<img width="1440" alt="Screenshot 2020-06-03 at 16 13 09" src="https://user-images.githubusercontent.com/25057618/83628167-b4d8eb80-a5b5-11ea-9803-9805dc8eb8a9.png">

Espresso Test Passes
<img width="1440" alt="Screenshot 2020-06-03 at 16 15 58" src="https://user-images.githubusercontent.com/25057618/83628112-9d99fe00-a5b5-11ea-98a3-785e415c463e.png">

`ProfilePictureActivityTest `
Robolectric Test Passes
<img width="1440" alt="Screenshot 2020-06-03 at 16 30 28" src="https://user-images.githubusercontent.com/25057618/83629454-bdcabc80-a5b7-11ea-8b57-2d5a5cbdd9d3.png">

Espresso Test Passes
<img width="1440" alt="Screenshot 2020-06-03 at 16 31 03" src="https://user-images.githubusercontent.com/25057618/83629433-b4d9eb00-a5b7-11ea-8434-9fd5b64caed5.png">

`InputInteractionViewTestActivityTest`
Robolectric Test Passes
<img width="1440" alt="Screenshot 2020-06-03 at 16 36 04" src="https://user-images.githubusercontent.com/25057618/83630176-03d45000-a5b9-11ea-9d7a-e55ca1e1d058.png">

Espresso Test Passes
<img width="1440" alt="Screenshot 2020-06-03 at 16 38 44" src="https://user-images.githubusercontent.com/25057618/83630149-f9b25180-a5b8-11ea-8a00-db60330c2e40.png">

**Comment by veena14cs** [comment]
> @veena14cs `MyDownloadsFragmentTest`, `ProfilePictureActivityTest` and `InputInteractionViewTestActivityTest` is good. Could we mark them done on the list?
> 

MyDownloadFragmentTest we cannot mark it as done. As there is still actual implementation has to be done. @rt4914 please confirm on this.

`InputInteractionViewTestActivityTest`  we can mark it as done but just want to confirm with @nikitamarysolomanpvt and @BenHenning if we have to wait for number with units implementation and testcases.

We can mark `ProfilePictureActivityTest` as done.

**Comment by rt4914** [comment]
@veena14cs @anandwana001 You can mark `MyDownloadFragmentTest` as done. Though the implementation is pending but it won't come till next month. And we will update everything accordingly.

**Comment by rt4914** [comment]
@veena14cs @anandwana001 Regarding `InputInteractionViewTestActivityTest ` also we can mark it has done. The main point is that we are marking done to all those files which are passing on roboelectric as well as espresso. And if we introduce new test cases we will make sure that those test cases pass on both.

**Comment by veena14cs** [comment]
> @veena14cs @anandwana001 You can mark `MyDownloadFragmentTest` as done. Though the implementation is pending but it won't come till next month. And we will update everything accordingly.

Thanks for confirming @rt4914 I will mark it as done then.

**Comment by veena14cs** [comment]
> @veena14cs @anandwana001 Regarding `InputInteractionViewTestActivityTest ` also we can mark it has done. The main point is that we are marking done to all those files which are passing on roboelectric as well as espresso. And if we introduce new test cases we will make sure that those test cases pass on both.

In that case will move this tests then. Thanks @rt4914

**Comment by BenHenning** [comment]
FYI that I'm taking StateFragmentTest in #1630. To everyone else: I suggest referencing this PR for future efforts since it should significantly simplify bringing other tests over to the non-flaky app module workflow.

**Comment by BenHenning** [comment]
It seems like SplashActivityTest was removed or was never added--it's going to be re-added to CI as part of #1397.

**Comment by anandwana001** [comment]
### Robolectric 
<img width="900" height="500" alt="Screenshot 2020-09-07 at 11 00 40" src="https://user-images.githubusercontent.com/25057618/92351502-83fadc00-f0f9-11ea-86a0-91a489cb5764.png">

### Espresso
<img width="900" height="500" alt="Screenshot 2020-09-07 at 10 57 35" src="https://user-images.githubusercontent.com/25057618/92351519-8e1cda80-f0f9-11ea-8e05-abfd0d1e5fc5.png">

**Comment by anandwana001** [comment]
HtmlParserTest is working perfectly, marking it down in the list. 
<img width="900" alt="Screenshot 2020-09-14 at 16 02 41" src="https://user-images.githubusercontent.com/25057618/93075674-c1380e80-f6a3-11ea-8bad-fa6f2bee70c8.png">
<img width="900" alt="Screenshot 2020-09-14 at 16 00 49" src="https://user-images.githubusercontent.com/25057618/93075689-c7c68600-f6a3-11ea-8030-a93c15f5cc86.png">

**Comment by anandwana001** [comment]
If anyone tryto fix **`OnboardingFragmentTest `**, the `skip_text_view` click is not working in the robolectric test and part from that `testOnboardingFragment_checkSlide3_clickGetStartedButton_opensProfileActivity` can be simply fixed by putting `scrollTo()` at `onView(withId(R.id.get_started_button)).perform(scrollTo(), click())`

**Comment by anandwana001** [comment]
`BindableAdapterTest` working perfectly on both espresso and robolectric, marking it correct.

<img width="900" alt="Screenshot 2020-09-14 at 22 14 48" src="https://user-images.githubusercontent.com/25057618/93114262-037b4300-f6d8-11ea-9bee-9de5fa375bc7.png">



<img width="900" alt="Screenshot 2020-09-14 at 22 17 22" src="https://user-images.githubusercontent.com/25057618/93114723-90260100-f6d8-11ea-80a0-9872a0d11f66.png">

**Comment by anandwana001** [comment]
`RevisionCardFragmentTest` works fine

### Robolectric
<img width="1440" alt="Screenshot 2020-09-20 at 19 01 57" src="https://user-images.githubusercontent.com/25057618/93712503-e8934d80-fb73-11ea-9650-c21ed848f1d8.png">

### Espresso
<img width="1440" alt="Screenshot 2020-09-20 at 19 04 22" src="https://user-images.githubusercontent.com/25057618/93712535-2001fa00-fb74-11ea-9bfd-905f5700ab74.png">

**Comment by prayutsu** [comment]
I am taking up ```OptionsFragmentTest```

**Comment by Sarthak2601** [comment]
@rt4914 Since we are going to take this issue under SLoP, can we freeze its contributions for now ? It might lead to confusion otherwise. 
@prayutsu Please wait on Rajat to reply. 
/cc- @anandwana001

**Comment by anandwana001** [comment]
> @rt4914 Since we are going to take this issue under SLoP, can we freeze its contributions for now ? It might lead to confusion otherwise. 
> @prayutsu Please wait on Rajat to reply. 
> /cc- @anandwana001 

Sounds good

**Comment by Sarthak2601** [comment]
@BenHenning should we include `StateFragementTest` in the tests for the SLoP participants to fix or you're working on it ?

**Comment by BenHenning** [comment]
@Sarthak2601 StateFragmentTest is fixed & running (as of #1630). Thanks for checking.

That being said, #388 & https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt#L206 might be a great set of tests that could be split into starter issues for SLoP. Ditto for question player (#503). #1273 also needs to be done (hints & solutions tests for questions).

@rt4914 we talked about not adding missing test cases, but much of the above is already figured out & immediately high impact since it covers the core learner experiences. It seems worth finishing that work.

Another unrelated area of work that I thought of: creating common test dependencies for matchers, shared modules, and shared utility functions (especially the ones used between state & question players).

**Comment by rt4914** [comment]
Closing this issue as we have created separate issues:
#1868 #1869 #1870 #1871 #1872 #1873 #1874 #1877 #1878
